src/core/Rule.ml

```diff
@@ -336,6 +321,72 @@ type error =
 (* can't use Error because it's used for severity *)
 exception Err of error
 
+(*****************************************************************************)
+(* Lazy parsing *)
+(*****************************************************************************)
+
+type rule_result = Rule of rule | InvalidRuleError of invalid_rule_error
```

Comment:
Did you try instead to just make the parsing of the pattern lazy? so have Xpattern.t defined as lazy?

Reply:
That way we don't need define a new rule_result type.

Reply:
I did try that first but it seemed like a bigger change. At least, the way I tried it, I was changing the type of rule, but this meant that everywhere rules were processed I had to double check whether the pattern was lazy or not. I liked having an invariant that once the rules were passed down for matching, they were parsed in full.

Reply:
I would expect a smaller change. Just change the Xpattern.t in Rule.ml to Xpattern.t lazy.

Reply:
But then what about everywhere the xpattern is used? Maybe I'm not understanding something about the type

Reply:
I thought there would not be many places where we use Xpattern.t, but indeed at those places
you would need to use Lazy.force or do like Nat does and use (lazy X) in the pattern to force the evaluation
directly from the pattern.

Reply:
Ah yeah there are a decent number of places. There's Analyze_rule, Translate_rule, Check_rule, and then at least four places in Match_search_mode. I didn't follow it all the way down the chain to see how many places I would need to change for sure, so there could be more. Given how many places we work with patterns I felt that this was nicer. I don't feel very strongly though, so I can change it if you do